Trello 19

ZiyedB opened the pull request:
Draft PR ,only look at the last commit. We need to get the other PRs reviewed and approved to merge them. 

Basic so far : 
![ezgif-2-5879bce76a3b](https://user-images.githubusercontent.com/6785920/111079371-a24ef380-84f1-11eb-91f9-5ec180367d03.gif)

Will add the :
- Add
- Enable/Disable
- Remove
- Edit
- Save

ZiyedB committed:
feat(card): add card component + extend tailwind instead of overwrite

ZiyedB committed:
feat(table): add a simple table component with tailwind
TRELLO-24 #Done

ZiyedB committed:
pr feedback

ZiyedB committed:
pr feedback

ZiyedB committed:
feat(tracked-contracts): ability to search through the chain to select and view the tracked contracts

TRELLO-19 #Done

ZiyedB commented on src/pages/tracked-contracts/mock-data.ts:
```diff
@@ -0,0 +1,33 @@
+// We can remove this when we can hook to the BE and get that data it's just for getting the UX ready for it
+const mockDataChain = [
```
Just mocking the sample we got from Jure

Frenkiee commented on src/pages/tracked-contracts/TrackedContractsList.tsx:
```diff
@@ -0,0 +1,96 @@
+          {chain.name}
+        </div>
+      );
+    });
```
A bit nicer:
```
 const allChains = chains
    .filter((chain) => !searchChain)
    .filter((chain) => chain.name.includes(searchChain.trim().toLocaleLowerCase()))
    .map((chain) => (
        <div
          className={`focus:outline-none py-4 capitalize cursor-pointer ${
            chain.name === activeChain?.name ? 'font-bold text-primary' : ''
          }`}
          role="button"
          onClick={() => setActiveChain(chain)}
          onKeyPress={() => setActiveChain(chain)}
          tabIndex={0}
        >
          {chain.name}
        </div>
      ));
```

Frenkiee replied:
Also, I would try to move out the div view and present it as a component

ZiyedB replied:
The filter like that wouldn't actually work but yeah simplifying it and moving it to a separate component.

Frenkiee commented on src/pages/tracked-contracts/TrackedContractsList.tsx:
```diff
@@ -0,0 +1,96 @@
+          return contract;
+        })
+      : 'No Tracked Contracts';
+
```
This should also work
```
const activeContracts =
    activeChain && activeChain.tracked_contracts.length
      ? activeChain.tracked_contracts
      : 'No Tracked Contracts';
```

Frenkiee commented on src/pages/tracked-contracts/TrackedContractsList.tsx:
```diff
@@ -0,0 +1,96 @@
+  project: string;
+  url: string;
+  credentials: {
+    // TODO: Ask potentially BE to have structure camelCase in case or we can just disable the warning in eslint
```
I've changed eslint settings in Query Builder to display a warning

ZiyedB committed:
pr feedback

ZiyedB commented on src/pages/tracked-contracts/TrackedChain.tsx:
```diff
@@ -0,0 +1,23 @@
+  onClick: () => void;
+}
+
+const TrackedChain: React.FC<TrackedChainProps> = ({ name, active, onClick }) => {
```
Moved it here, in the same folder, I don't think it's gonna be used anywhere else so no need to put it inside components

ZiyedB commented on src/pages/tracked-contracts/TrackedContractsList.tsx:
```diff
@@ -28,38 +29,15 @@ const TrackedContractsList: React.FC<{}> = () => {
-        </div>
-      );
-    });
+    .filter((chain) => chain.name.includes(searchChain.trim().toLocaleLowerCase()) || activeChain?.name === chain.name)
```
Just this filter works actually no need to check for if defined or null. Also added the fact that if a chan is already active you might still want to see it despite searching

ZiyedB committed:
basic ui implementation

ZiyedB committed:
hook with the BE

ZiyedB commented:
<img width="740" alt="Screenshot 2021-04-01 at 16 14 24" src="https://user-images.githubusercontent.com/6785920/113315676-714e2b80-9305-11eb-9c16-0d597c769823.png">
<img width="747" alt="Screenshot 2021-04-01 at 16 14 21" src="https://user-images.githubusercontent.com/6785920/113315681-727f5880-9305-11eb-8fef-429b95610ace.png">
<img width="1103" alt="Screenshot 2021-04-01 at 16 07 20" src="https://user-images.githubusercontent.com/6785920/113315682-7317ef00-9305-11eb-9db4-2097562398b2.png">

ZiyedB commented on src/api/contracts.ts:
```diff
@@ -0,0 +1,17 @@
+/* eslint-disable @typescript-eslint/explicit-module-boundary-types */
+
+// Implementation : https://github.com/paralink-network/paralink-node/blob/master/src/api/contracts.py
+const getAllContracts = () => axiosInstance.get<Contract[]>(`${contractsApi}`);
```
FYI @frisitano @jbargu I've tried some of the BE Apis and had some feedback :  

1. https://github.com/paralink-network/paralink-node/blob/master/src/api/contracts.py#L47 POST & PUT should return the contract response instead of just OK ( we'd need the latest from the DB, especially in the POST to have the ID without refreshing the page or refetching all of the contracts ) 


2. Error when trying to update the status of an address, we might want to fix that
<img width="577" alt="Screenshot 2021-04-01 at 15 52 43" src="https://user-images.githubusercontent.com/6785920/113316140-eb7eb000-9305-11eb-9b13-5eab6a4222cd.png">

3. Should we be able to edit an address after it was created or saved ?  cause we don't have a "setAddress" or a modify endpoint. Also if we were to add one, let's just do it as "updateContract" endpoint and remove the status one.

4. Error when deleting. It actually works, but return a 500 as there is an error after deleting : 
<img width="860" alt="Screenshot 2021-04-01 at 15 48 21" src="https://user-images.githubusercontent.com/6785920/113316143-ec174680-9305-11eb-9f0b-1dcf06397678.png">

frisitano replied:
Hi @ZiyedB

I have made some modifications to the API contract on this development branch - [refactor_models ](https://github.com/frisitano/paralink-node/tree/refactor_models).

This refactor modifies the API such that contracts are identified by a unique (chain, address) tuple.  

1) Under this new API you should not need any additional information returned by the API as the contract id has been removed.  

2) I am not experiencing the same issue as you.  I suspect your `.env` file may be referencing the celery broker incorrectly.  Could you ensure it contains `CELERY_BROKER_URL=amqp://guest:guest@rabbitmq:5672` please?

3) No we should not be able to update the address.  A different address would represent a different contract.

4) Fixed.

Could you take a look at the API and let me know your feedback.  I will hold fire on raising a PR to see if we need to make any modifications to the API for the front end.  In any case I still need to develop some tests for the REST endpoints.

frisitano replied:
I forgot to mention that you will have to delete the docker volume associated with the paralink Postgres container before executing the dev branch above.

ZiyedB commented on src/interfaces/Chain.ts:
```diff
@@ -0,0 +1,8 @@
+import { Contract } from './Contract';
```
FYI @Frenkiee I moved things inside the interfaces folder, I think it might be much easier to share the interfaces across things ( apis / components / utils etc. ) avoiding some circular deps later on

ZiyedB commented on src/pages/tracked-contracts/TrackedContract.tsx:
```diff
@@ -1,59 +1,82 @@
 import React, { useEffect, useState } from 'react';
 import { FaCheck, FaTimes, FaTrash } from 'react-icons/fa';
+import ContractsApi from '../../api/contracts';
```
I've done it a slightly differently from you @Frenkiee , I like that format of just importing as named import and then calling the function.

The then / catch & everything else is done directly from the component.

ZiyedB commented on src/pages/tracked-contracts/TrackedContractsList.tsx:
```diff
@@ -33,46 +18,48 @@ const TrackedContractsList: React.FC<{}> = () => {
-
   useEffect(() => {
-    setChains(mockDataChain as Chain[]);
+    ChainsApi.getAllChains().then((res: any) => {
```
If we can get another chain added

ZiyedB commented on src/pages/tracked-contracts/TrackedContract.tsx:
```diff
@@ -1,59 +1,82 @@
 
   // Save the contract
   const saveClick = (): void => {
     // save contract as it is
```
Before saving we would check that the address is valid or not using some utils. Ticket to come about adding a util to check input address based on the chain

Good example from Jure : https://polkadot.js.org/docs/util-crypto/examples/validate-address/

frisitano replied:
This is handled in the backend [here](https://github.com/paralink-network/paralink-node/blob/master/src/models/contract.py#L83).  You should look out for the [error](https://github.com/paralink-network/paralink-node/blob/master/src/api/contracts.py#L61) when calling the `create_contract` endpoint.

ZiyedB commented on src/pages/tracked-contracts/TrackedContractsList.tsx:
```diff
@@ -33,46 +18,48 @@ const TrackedContractsList: React.FC<{}> = () => {
   const allChains = chains
     .filter((chain) => chain.name.includes(searchChain.trim().toLocaleLowerCase()) || activeChain?.name === chain.name)
     .map((chain) => (
-      <TrackedChain active={chain.name === activeChain?.name} name={chain.name} onClick={() => setActiveChain(chain)} />
```
Thinking about it @jbargu  and your suggestion, I might remove the search and instead have a selector there

<img width="294" alt="Screenshot 2021-04-01 at 16 27 08" src="https://user-images.githubusercontent.com/6785920/113317302-28977200-9307-11eb-8678-0b3b054267e1.png">
Or we could have just one small selectors to select the type 'evm' | 'substrate' ( or both )

ZiyedB committed:
setup correclty the ref

ZiyedB committed:
update with latest model

ZiyedB commented:
FYI @frisitano I've worked on the updates. It's looking good, here is a small gif , you can also try it out. ( added the notifications while at it )

![ezgif com-gif-maker](https://user-images.githubusercontent.com/6785920/117569891-b2351f00-b0bf-11eb-9fc8-81a0fdbeabab.gif)

Small thing but not blocking we can address later, as I didn't yet added the "insights" about errors yet but it looks like we'd need a json format for the error. Right now it's an HTML format error.

<img width="627" alt="Screenshot 2021-05-09 at 12 12 01" src="https://user-images.githubusercontent.com/6785920/117569976-1c4dc400-b0c0-11eb-8554-8ab4c3ce3033.png">

frisitano commented:
Hi @ZiyedB I've tried out the UI and it looks great so far.  I have updated the error response such that it is json as required, please go ahead and pull the latest.  You can see [the source code](https://github.com/frisitano/paralink-node/blob/refactor_models/src/api/contracts.py#L61) as a reference.  I believe the code should be intuitive and self explanatory.  In the error scenario you can fetch the error message to display to the user from the "error" field in the json.

ZiyedB committed:
integrate with errors

ZiyedB commented:
Thanks @frisitano I've added the support for the error , it looks good : 
<img width="1017" alt="Screenshot 2021-05-17 at 22 57 42" src="https://user-images.githubusercontent.com/6785920/118562232-393f6280-b764-11eb-94ce-2604aeb964a7.png">
I've added an autoclose after 8s instead, but we can think of persistent error message later in case.

Also for later, I think we should add the following : 
- Errors for other endpoints
- Potentially just throwing an error ( as right now it's still considered as "response" ) ,for this no need yet, we could potentially try to add a response interceptor to throw an error on our end based on the result, I added a TODO for that, but not sure if it's doable with axios ( need to check )
